### ISSUE TITLE
[BUG] Unable to reproduce results on different UNIX machines with TimeSeriesPredictor (TFT and PatchTST)

### ISSUE THREAD

**Issue by valeju** [autogluon/autogluon#4392]
**Bug Report Checklist**

- [x] I provided code that demonstrates a minimal reproducible example.  
- [x] I confirmed bug exists on the latest mainline of AutoGluon via source install.  
- [x] I confirmed bug exists on the latest stable version of AutoGluon.  

**Describe the bug**

I am training and testing a TimeSeriesPredictor model using various sub-models. When I run my code on two different UNIX machine

* UNIX machine **1** has 32 cores and 64GB RAM (cf details below)
* UNIX machine **2** has 8 cores and 8GB RAM (cf details below)

I observed discrepancies in the results for the `TFT` and `PatchTST` models. Consequently, the WeightedEnsemble also produced differing results. The random seed used is the default (`123`). The data used to train and predict is the same.

Note that the UNIX machines don't have GPUs.
Note 2: If the code is ran several times on the same machine the final predictions are always identical.

**Expected behavior**

If my code runs in different UNIX machines, The final predictions must be the same regardless of the number of CPUs or RAM.

**To Reproduce**
Please see the attached notebook. It load the data, create custom known covariates, trains a TimeSeriesPredictor model and predict the values for different windows of the test sets.

<details>

```python
import pandas as pd
import numpy as np
from autogluon.timeseries import TimeSeriesDataFrame, TimeSeriesPredictor
from autogluon.timeseries.utils.forecast import get_forecast_horizon_index_ts_dataframe
from autogluon.timeseries.splitter import ExpandingWindowSplitter

known_covariates_names=["is_weekday", "hour", "is_market_open" ]
timestamp_col = 'timestamp'
item_ids = ['H1', 'H2', 'H3']
item_id_col = 'item_id'
prediction_length = 4
num_val_windows = 6
target_name = 'target'

def create_known_covariates(df, timestamp_col=None, dt=None):
    weekday_values = np.arange(0, 5, 1)
    working_hours_values = np.arange(8, 18, 1)
    assert (timestamp_col is not None) or (dt is not None), "Either `timestamp_col` or `dt` must be set."
    if timestamp_col is not None:
        datetime_properties = df[timestamp_col].dt
    elif dt is not None:
        datetime_properties = dt
        
    df["is_weekday"] = datetime_properties.weekday.isin(weekday_values)
    df["hour"] = datetime_properties.hour
    df["is_market_open"] = df['is_weekday'] & df['hour'].isin(working_hours_values)
    return df

train_df = pd.read_csv("https://autogluon.s3.amazonaws.com/datasets/timeseries/m4_hourly_subset/train.csv", parse_dates=[timestamp_col])
train_df = create_known_covariates(train_df, timestamp_col=timestamp_col)
train_df = train_df[train_df[item_id_col].isin(item_ids)]
train_data = TimeSeriesDataFrame.from_data_frame(train_df, id_column=item_id_col, timestamp_column=timestamp_col)

test_df = pd.read_csv("https://autogluon.s3.amazonaws.com/datasets/timeseries/m4_hourly_subset/test.csv", parse_dates=[timestamp_col])
test_df = test_df[test_df[item_id_col].isin(item_ids)]
test_df = create_known_covariates(test_df, timestamp_col=timestamp_col)
test_data = TimeSeriesDataFrame.from_data_frame(test_df, id_column=item_id_col, timestamp_column=timestamp_col)

predictor = TimeSeriesPredictor(
    prediction_length=prediction_length,
    path="bug_autogluon_pod_small",
    target=target_name,
    known_covariates_names=known_covariates_names,
    eval_metric="MASE"
)

# fit
predictor.fit(
    train_data,
    hyperparameters={
        "TemporalFusionTransformer": {"context_length": 16, "hidden_dim": 8, "max_epochs": 10},
        "AutoETS": {},
        "CrostonSBA": {},
        "SeasonalNaive": {},
        "DynamicOptimizedTheta": {},
        "AutoARIMA": {"max_ts_length": 512},
        "DLinear": {"hidden_dimension": 8, "max_epochs": 10},
        "PatchTST": {"max_epochs": 10}
    },
    num_val_windows=num_val_windows,
    refit_every_n_windows=2
)

# test
splitter = ExpandingWindowSplitter(prediction_length=prediction_length, num_val_windows=num_val_windows)
predictions = []
for window_idx, (train_split, val_split) in enumerate(splitter.split(test_data)):
    future_index = get_forecast_horizon_index_ts_dataframe(val_split, prediction_length=prediction_length)
    known_covariates_data = pd.DataFrame(index=future_index)
    known_covariates_data = create_known_covariates(known_covariates_data, dt=future_index.get_level_values("timestamp"))
    for m in predictor.model_names():
        preds = pd.DataFrame(predictor.predict(val_split, known_covariates_data, model=m, random_seed=123))
        preds['model'] = m
        predictions.append(preds.set_index('model', append=True))
predictions = pd.concat(predictions)

predictions.to_parquet("prediction_run_1.parquet")

## Please uncomment those lines to compare predictions from 2 different runs.
## load predictions from another run
# predictions_other_run = pd.read_parquet("predictions_other_run.parquet")  # <-- change the path here.
## Create the absolute difference dataframe between 2 runs
# abs_diff_df = (predictions_other_run - predictions).abs()
## Compute the mean absolute difference for each model and each item id.
# abs_diff_df.groupby(['item_id', 'model']).mean()
```
</details>

**Screenshots / Logs**

Below is a picture of the mean absolute error between predictions obtained with UNIX machine 1 and the predictions with UNIX machine 2. The MAE of the predictions must be 0, yet for `TFT`, `PatchTST`  and `WeightedEnsemble` the discrepancies are significant.
![image](https://github.com/user-attachments/assets/39c9cd11-1aec-49a0-b7b3-0a91ae1a6d24)

**Installed Versions**
#### Unix machine 1
<details>

```python
# Replace this code with the output of the following:
from autogluon.core.utils import show_versions
show_versions()
```
INSTALLED VERSIONS of UNIX Machine # 1
------------------
date                : 2024-08-15
time                : 08:16:26.062322
python              : 3.10.14.final.0
OS                  : Linux
OS-release          : 5.15.0-1066-azure
Version             : # 75-Ubuntu SMP Thu May 30 14:29:45 UTC 2024
machine             : x86_64
processor           : x86_64
num_cores           : 32
cpu_ram_mb          : 257926.23046875
cuda version        : None
num_gpus            : 0
gpu_ram_mb          : []
avail_disk_size_mb  : 655053

accelerate          : 0.21.0
autogluon           : None
autogluon.common    : 1.1.1
autogluon.core      : 1.1.1
autogluon.features  : 1.1.1
autogluon.tabular   : 1.1.1
autogluon.timeseries: 1.1.1
boto3               : 1.34.156
catboost            : 1.2.5
fastai              : None
gluonts             : 0.15.1
hyperopt            : 0.2.7
imodels             : None
joblib              : 1.4.2
lightgbm            : 3.3.5
lightning           : 2.3.3
matplotlib          : 3.9.1.post1
mlforecast          : 0.10.0
networkx            : 3.3
numpy               : 1.25.2
onnxruntime-gpu     : None
optimum             : None
optimum-intel       : None
orjson              : 3.10.6
pandas              : 2.0.3
psutil              : 5.9.8
pytorch-lightning   : 2.3.3
ray                 : 2.10.0
requests            : 2.32.3
scikit-learn        : 1.4.0
scikit-learn-intelex: None
scipy               : 1.12.0
setuptools          : 72.1.0
skl2onnx            : None
statsforecast       : 1.4.0
tabpfn              : None
tensorboard         : 2.17.0
torch               : 2.3.1
tqdm                : 4.66.5
transformers        : 4.40.2
utilsforecast       : 0.0.10
vowpalwabbit        : None
xgboost             : 1.7.6
</details>

#### Unix machine 2
<details>

```python
# Replace this code with the output of the following:
from autogluon.core.utils import show_versions
show_versions()
```
INSTALLED VERSIONS of UNIX Machine # 2
------------------
date                : 2024-08-15
time                : 10:31:44.659581
python              : 3.10.14.final.0
OS                  : Linux
OS-release          : 5.15.0-1052-azure
Version             : # 60-Ubuntu SMP Mon Nov 6 10:08:16 UTC 2023
machine             : x86_64
processor           : x86_64
num_cores           : 8
cpu_ram_mb          : 32094.55078125
cuda version        : None
num_gpus            : 0
gpu_ram_mb          : []
avail_disk_size_mb  : 655028

accelerate          : 0.21.0
autogluon           : None
autogluon.common    : 1.1.1
autogluon.core      : 1.1.1
autogluon.features  : 1.1.1
autogluon.tabular   : 1.1.1
autogluon.timeseries: 1.1.1
boto3               : 1.34.156
catboost            : 1.2.5
fastai              : None
gluonts             : 0.15.1
hyperopt            : 0.2.7
imodels             : None
joblib              : 1.4.2
lightgbm            : 3.3.5
lightning           : 2.3.3
matplotlib          : 3.9.1.post1
mlforecast          : 0.10.0
networkx            : 3.3
numpy               : 1.25.2
onnxruntime-gpu     : None
optimum             : None
optimum-intel       : None
orjson              : 3.10.6
pandas              : 2.0.3
psutil              : 5.9.8
pytorch-lightning   : 2.3.3
ray                 : 2.10.0
requests            : 2.32.3
scikit-learn        : 1.4.0
scikit-learn-intelex: None
scipy               : 1.12.0
setuptools          : 72.1.0
skl2onnx            : None
statsforecast       : 1.4.0
tabpfn              : None
tensorboard         : 2.17.0
torch               : 2.3.1
tqdm                : 4.66.5
transformers        : 4.40.2
utilsforecast       : 0.0.10
vowpalwabbit        : None
xgboost             : 1.7.6
</details>

**Comment by shchur** [comment]
Short update: I wasn't able to find a solution to this problem before the AutoGluon 1.2 release, so I'm moving this to the 1.3 milestone. 

It is possible that we will be able to fix this problem on the GluonTS side, in which case the fix will be available in the next GluonTS release (e.g., 0.16.1) that will automatically propagate to AutoGluon since our version cap is `gluonts>=0.15,<0.17`.

**Comment by valeju** [comment]
> Short update: I wasn't able to find a solution to this problem before the AutoGluon 1.2 release, so I'm moving this to the 1.3 milestone.
> 
> It is possible that we will be able to fix this problem on the GluonTS side, in which case the fix will be available in the next GluonTS release (e.g., 0.16.1) that will automatically propagate to AutoGluon since our version cap is `gluonts>=0.15,<0.17`.

Thanks for the update.
